flink-core/src/main/java/org/apache/flink/api/common/typeutils/TypeSerializerSnapshot.java

```diff
@@ -124,11 +124,40 @@ void readSnapshot(int readVersion, DataInputView in, ClassLoader userCodeClassLo
      * program's serializer re-serializes the data, thus converting the format during the restore
      * operation.
      *
+     * @deprecated This method has been replaced by {@link TypeSerializerSnapshot
+     *     #resolveSchemaCompatibility(TypeSerializerSnapshot)}.
      * @param newSerializer the new serializer to check.
      * @return the serializer compatibility result.
      */
-    TypeSerializerSchemaCompatibility<T> resolveSchemaCompatibility(
-            TypeSerializer<T> newSerializer);
+    @Deprecated
+    default TypeSerializerSchemaCompatibility<T> resolveSchemaCompatibility(
+            TypeSerializer<T> newSerializer) {
+        return newSerializer.snapshotConfiguration().resolveSchemaCompatibility(this);
+    }
+
+    /**
+     * Checks current serializer's compatibility to read data written by the prior serializer.
+     *
+     * <p>When a checkpoint/savepoint is restored, this method checks whether the serialization
+     * format of the data in the checkpoint/savepoint is compatible for the format of the serializer
+     * used by the program that restores the checkpoint/savepoint. The outcome can be that the
+     * serialization format is compatible, that the program's serializer needs to reconfigure itself
+     * (meaning to incorporate some information from the TypeSerializerSnapshot to be compatible),
+     * that the format is outright incompatible, or that a migration needed. In the latter case, the
+     * TypeSerializerSnapshot produces a serializer to deserialize the data, and the restoring
+     * program's serializer re-serializes the data, thus converting the format during the restore
+     * operation.
+     *
+     * <p>This method must be implemented to clarify the compatibility. See FLIP-263 for more
+     * details.
+     *
+     * @param oldSerializerSnapshot the old serializer snapshot to check.
+     * @return the serializer compatibility result.
+     */
+    default TypeSerializerSchemaCompatibility<T> resolveSchemaCompatibility(
+            TypeSerializerSnapshot<T> oldSerializerSnapshot) {
+        return oldSerializerSnapshot.resolveSchemaCompatibility(restoreSerializer());
```

Comment:
I'm not sure should we add a default implementation for the new `resolveSchemaCompatibility` method. I'm worried that providing a default implementation may not work well in some scenarios. We have 2 `resolveSchemaCompatibility` methods for now, and they call each other by default. This behavior may be a bug, for example: 
- The `TypeSerializerSnapshot` of `data1` has old `resolveSchemaCompatibility(TypeSerializer)` method.
- The `TypeSerializerSnapshot` of `data2` has new `resolveSchemaCompatibility(TypeSerializerSnapshot)` method.

When flink users upgrade the data from data2 to data1, and flink will call the `resolveSchemaCompatibility(TypeSerializerSnapshot)` of data1, it will call the default implementation, so data1 will call the data2's `resolveSchemaCompatibility(TypeSerializer)`, and then it will call the default implementation. So call `resolveSchemaCompatibility(TypeSerializerSnapshot)` of data1 again.

It's a dead loop.

Based on this, I suggest all types inside of flink project should implement new `resolveSchemaCompatibility(TypeSerializerSnapshot)`. Currently, the `EitherSerializerSnapshot` and `GenericArraySerializerConfigSnapshot` don't implement it. WDYT?

Can we not provide a default implementation for the new `resolveSchemaCompatibility(TypeSerializerSnapshot)`? The default implementation is provided because the user may have a custom serializer, right?

Reply:
Thanks for the detailed review.
IIUC, Only If users has implemented one of the methods (old one or new one), it should work well in current PR.
Sorry maybe I missed your point. The case you described is an nested serializer, right ?
Whichever method `data1` and `data2` implement, the call should be onesided, otherwise the logic should be error even if we don't have this pr.

The reason why we provide two default methods is that users could not use mixed migrated method and non-migrated method as you could saw the discussion [#21635(comment)](https://github.com/apache/flink/pull/21635#discussion_r1086625857)

Just cc @Zakelly as we also talked about this problem offline.

Reply:
After thinking more about it, I'm wondering if we don't need a compatibility implementation for the old `resolveSchemaCompatibility(TypeSerializer)` method.

IIUC, Flink just calls the new `resolveSchemaCompatibility(TypeSerializerSnapshot)` method in the future, right?

If so, we can add a default implementation for the old `resolveSchemaCompatibility(TypeSerializer)` method, the default implementation throwes `UnsupportedException`. We need to add a default implementation because users or developers don't need to implement it in the future.

For the new `resolveSchemaCompatibility(TypeSerializerSnapshot)` method, you current implementation is fine. All old implementation class can be supported if `resolveSchemaCompatibility(TypeSerializerSnapshot)` calls the old one.

BTW, the code comment should guide users or developers to implement the new `resolveSchemaCompatibility(TypeSerializerSnapshot)` method in the future.

WDYT?

Reply:
> IIUC, Flink just calls the new resolveSchemaCompatibility(TypeSerializerSnapshot) method in the future, right?

Yes, Flink could make sure this.
But IIUC, `TypeSerializerSnapshot` and `TypeSerializer` are public API which means they maybe also used in the user code. Then users still have to modify their codes which may break changes.

> BTW, the code comment should guide users or developers to implement the new resolveSchemaCompatibility(TypeSerializerSnapshot) method in the future.

Thanks for the advice. That's a good idea.

Reply:
>  `TypeSerializerSnapshot and TypeSerializer` are public API which means they maybe also used in the user code. Then users still have to modify their codes which may break changes.

Thanks for the clarification!